src/adc.rs

```diff
@@ -0,0 +1,534 @@
+);
+
+#[cfg(any(feature = "stm32f303x6", feature = "stm32f303x8"))]
+adc_pins!(ADC1,
+    gpiob::PB0<Analog> => 11_u8,
+    gpiob::PB1<Analog> => 12_u8,
+    gpiob::PB13<Analog> => 13_u8,
+);
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC1,
+    gpiof::PF4<Analog> => 5_u8,
+    gpiof::PF2<Analog> => 10_u8,
+);
+
+// # ADC2 Pin/Channel mapping
+// ## f303
+
+#[cfg(feature = "stm32f303")]
+adc_pins!(ADC2,
+    gpioa::PA4<Analog> => 1_u8,
+    gpioa::PA5<Analog> => 2_u8,
+    gpioa::PA6<Analog> => 3_u8,
+    gpioa::PA7<Analog> => 4_u8,
+    gpioc::PC4<Analog> => 5_u8,
+    gpioc::PC5<Analog> => 11_u8,
+    gpiob::PB2<Analog> => 12_u8,
+    gpioc::PC0<Analog> => 6_u8,
+    gpioc::PC1<Analog> => 7_u8,
+    gpioc::PC2<Analog> => 8_u8,
+    gpioc::PC3<Analog> => 9_u8,
+);
+
+#[cfg(any(feature = "stm32f303x6", feature = "stm32f303x8"))]
+adc_pins!(ADC2,
+    gpiob::PB12<Analog> => 13_u8,
+    gpiob::PB14<Analog> => 14_u8,
+    gpiob::PB15<Analog> => 15_u8,
+);
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC2,
+    gpiof::PF2<Analog> => 10_u8,
+);
+
+// # ADC3 Pin/Channel mapping
+// ## f303
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC3,
+    gpiob::PB1<Analog> => 1_u8,
+    gpioe::PE9<Analog> => 2_u8,
+    gpioe::PE13<Analog> => 3_u8,
+    gpiob::PB13<Analog> => 5_u8,
+    gpiob::PB0<Analog> => 12_u8,
+    gpioe::PE7<Analog> => 13_u8,
+    gpioe::PE10<Analog> => 14_u8,
+    gpioe::PE11<Analog> => 15_u8,
+    gpioe::PE12<Analog> => 16_u8,
+    // Shared channels (i.e. ADC34_INx)
+    gpioe::PE8<Analog> => 6_u8,
+    gpiod::PD10<Analog> => 7_u8,
+    gpiod::PD11<Analog> => 8_u8,
+    gpiod::PD12<Analog> => 9_u8,
+    gpiod::PD13<Analog> => 10_u8,
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+// # ADC4 Pin/Channel mapping
+// ## f303
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC4,
+    gpioe::PE14<Analog> => 1_u8,
+    gpioe::PE15<Analog> => 2_u8,
+    gpiob::PB12<Analog> => 3_u8,
+    gpiob::PB14<Analog> => 4_u8,
+    gpiob::PB15<Analog> => 5_u8,
+    gpiob::PB8<Analog> => 12_u8,
+    gpiob::PB9<Analog> => 13_u8,
+    // Shared channels (i.e. ADC34_INx)
+    gpioe::PE8<Analog> => 6_u8,
+    gpiod::PD10<Analog> => 7_u8,
+    gpiod::PD11<Analog> => 8_u8,
+    gpiod::PD12<Analog> => 9_u8,
+    gpiod::PD13<Analog> => 10_u8,
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+/// Abstract implementation of ADC functionality
+///
+/// Do not use directly. See adc12_hal for a applicable Macro.
+/// TODO: Extend/generalize beyond f303
+macro_rules! adc_hal {
+    ($(
+            $ADC:ident: ($init:ident, $ADC_COMMON:ident),
```

Comment:
```suggestion
            $ADC:ident: ($adcx:ident, $ADC_COMMON:ident),
```
This would be more consistent with other macro definitions, see:

https://github.com/stm32-rs/stm32f3xx-hal/blob/7e0d7e67617fdb12174f76872a00bd2f54a6eb01/src/gpio.rs#L103